waddrmgr/manager.go

```diff
@@ -1360,15 +1363,31 @@ func (m *Manager) Unlock(passphrase []byte) error {
 	return nil
 }
 
-// MarkUsed updates the used flag for the provided address id.
-func (m *Manager) MarkUsed(addressID []byte) error {
+// fetchUsed returns true if the provided address id was flagged used.
+func (m *Manager) fetchUsed(addressID []byte) (bool, error) {
+	var used bool
+	err := m.namespace.View(func(tx walletdb.Tx) error {
+		used = fetchAddressUsed(tx, addressID)
+		return nil
+	})
+	return used, err
+}
+
+// MarkUsed updates the used flag for the provided address.
+func (m *Manager) MarkUsed(address btcutil.Address) error {
+	addressID := address.ScriptAddress()
 	err := m.namespace.Update(func(tx walletdb.Tx) error {
 		return markAddressUsed(tx, addressID)
 	})
 	if err != nil {
 		return maybeConvertDbError(err)
 	}
-	// 'used' flag has become stale so remove key from cache
+	// Clear caches which might have stale entries for used addresses
+	account, err := m.AddrAccount(address)
+	if err != nil {
+		return err
+	}
+	delete(m.acctInfo, account)
```

Comment:
@jrick This is a bit of a hack since `acctInfo` caches the derived last (ex/in) addresses. I guess a better solution would be to stop deriving addrs and always load them from the db.